File: Leetcode/Cpp/#70_Climbing_Stairs.cpp
// Date last reviewed:
// 4/9/2025
// =======================================================
class Solution {
public:
    int climbStairs(int n) {
        int steps[4];
        steps[1] = 1;
        steps[2] = 2;
        for(int i = 3; i < n+1; i++) {
            steps[i&3] = steps[(i-1)&3] + steps[(i-2)&3];
        }
        return steps[n&3];
    }
};

File: Leetcode/Python/#70_Climbing_Stairs.py
class Solution:
    def climbStairs(self, n: int) -> int:
        if n <= 2:
            return n
        ways = [1, 2]
        for i in range(2, n):
            ways[i&1] = ways[0] + ways[1]
        return ways[(n-1)&1]

class Solution:
    def climbStairs(self, n: int) -> int:
        if n < 3:
            return n
        dp = [1,2]
        for i in range(2,n+1):
            dp[i%2] = dp[0] + dp[1]
        return dp[(n+1)%2]
